src/lib.rs

```diff
@@ -299,33 +423,108 @@ pub unsafe extern "C" fn fd_seek(
     whence: Whence,
     newoffset: *mut Filesize,
 ) -> Errno {
-    unreachable()
+    match Descriptor::get(fd) {
+        Descriptor::File(file) => {
+            // It's ok to cast these indices; the WASI API will fail if
+            // the resulting values are out of range.
+            let from = match whence {
+                WHENCE_SET => wasi_filesystem::SeekFrom::Set(offset as _),
+                WHENCE_CUR => wasi_filesystem::SeekFrom::Cur(offset),
+                WHENCE_END => wasi_filesystem::SeekFrom::End(offset as _),
+                _ => return ERRNO_INVAL,
+            };
+            match file.fd.seek(from) {
+                Ok(result) => {
+                    *newoffset = result;
+                    ERRNO_SUCCESS
+                }
+                Err(err) => errno_from_wasi_filesystem(err),
+            }
+        }
+        Descriptor::Log => ERRNO_SPIPE,
+        Descriptor::Closed => ERRNO_BADF,
+    }
 }
 
 /// Synchronize the data and metadata of a file to disk.
 /// Note: This is similar to `fsync` in POSIX.
 #[no_mangle]
 pub unsafe extern "C" fn fd_sync(fd: Fd) -> Errno {
-    unreachable()
+    match Descriptor::get(fd) {
+        Descriptor::File(file) => match file.fd.sync() {
+            Ok(()) => ERRNO_SUCCESS,
+            Err(err) => errno_from_wasi_filesystem(err),
+        },
+        Descriptor::Log => ERRNO_INVAL,
+        Descriptor::Closed => ERRNO_BADF,
+    }
 }
 
 /// Return the current offset of a file descriptor.
 /// Note: This is similar to `lseek(fd, 0, SEEK_CUR)` in POSIX.
 #[no_mangle]
 pub unsafe extern "C" fn fd_tell(fd: Fd, offset: *mut Filesize) -> Errno {
-    unreachable()
+    match Descriptor::get(fd) {
+        Descriptor::File(file) => match file.fd.tell() {
+            Ok(result) => {
+                *offset = result;
+                ERRNO_SUCCESS
+            }
+            Err(err) => errno_from_wasi_filesystem(err),
+        },
+        Descriptor::Log => ERRNO_SPIPE,
+        Descriptor::Closed => ERRNO_BADF,
+    }
 }
 
 /// Write to a file descriptor.
 /// Note: This is similar to `writev` in POSIX.
 #[no_mangle]
 pub unsafe extern "C" fn fd_write(
     fd: Fd,
-    iovs_ptr: *const Ciovec,
-    iovs_len: usize,
+    mut iovs_ptr: *const Ciovec,
+    mut iovs_len: usize,
     nwritten: *mut Size,
 ) -> Errno {
-    unreachable()
+    // Advance to the first non-empty buffer.
+    while iovs_len != 0 && (*iovs_ptr).buf_len == 0 {
+        iovs_ptr = iovs_ptr.add(1);
+        iovs_len -= 1;
+    }
+    if iovs_len == 0 {
+        *nwritten = 0;
+        return ERRNO_SUCCESS;
+    }
+
+    let ptr = (*iovs_ptr).buf;
+    let len = (*iovs_ptr).buf_len;
+
+    match Descriptor::get(fd) {
+        Descriptor::File(file) => {
+            register_buffer(ptr as *mut _, len);
```

Comment:
I think this may not be necessary since `cabi_realloc` won't be called during the return value?

Reply:
Yes, that was a copy+pasto.